### ISSUE TITLE
Solve ImapConnection.java reuse issues

### ISSUE THREAD

**Issue by philipwhiuk** [thunderbird/thunderbird-android#1297]
### Expected behaviour

Should be able to re-authenticate with Gmail using a supported authentication method
### Actual behaviour

Authentication fails.
### Steps to reproduce
1. Set-up a Gmail IMAP account
2. Timeout the connections
3. See failure to connect
### Environment

K-9 Mail version: 97ea67d

Android version: 6.0.1

Account type (IMAP, POP3, WebDAV/Exchange): IMAP
### Comments

Gmail returns different capabilities pre and post-auth. Specifically, post-auth it returns a whole bunch of list and compression capabilities and pre-auth it returns a whole bunch of AUTH related capabilities.

Gmail reports different capabilities pre and post-auth and we just replace the capabilities. We only request capabilities if we have none.

```
04-17 20:14:47.614 30316-30353/com.fsck.k9.debug V/k9: conn176319345<<<#null# [CAPABILITY, IMAP4rev1, UNSELECT, IDLE, NAMESPACE, QUOTA, ID, XLIST, CHILDREN, X-GM-EXT-1, XYZZY, SASL-IR, AUTH=XOAUTH2, AUTH=PLAIN, AUTH=PLAIN-CLIENTTOKEN, AUTH=OAUTHBEARER, AUTH=XOAUTH]
04-17 20:14:49.453 30316-30353/com.fsck.k9.debug V/k9: conn176319345<<<#null# [CAPABILITY, IMAP4rev1, UNSELECT, IDLE, NAMESPACE, QUOTA, ID, XLIST, CHILDREN, X-GM-EXT-1, UIDPLUS, COMPRESS=DEFLATE, ENABLE, MOVE, CONDSTORE, ESEARCH, UTF8=ACCEPT, APPENDLIMIT=35882577, LIST-EXTENDED, LIST-STATUS]
```

```
04-17 20:13:59.469 30316-30381/com.fsck.k9.debug E/k9: Failed to login, closing connection for conn207765272
04-17 20:13:59.529 30316-30381/com.fsck.k9.debug E/k9: Got exception while idling for Gmail :Drafts/Thread-3978/conn207765272
                                                       com.fsck.k9.mail.MessagingException: Server doesn't support SASL XOAUTH2
```

Note that in the example I'm using SASL XOAUTH2 (with changes for #1295 )  but I have no reason to believe the behaviour would be different - GMail drops all auth privileges in the post auth capability list.

Because we don't expect this behaviour we just replace our capability listing with the one we get post-auth then if we close the connection when we reconnect we think they don't support the chosen method.

I'm not sure who isn't following the spec yet, them or us. You're allowed to change capability provision after STARTTLS (which we handle). But we don't handle different capabilities pre- and post-auth.

Anyway, even if they are breaking expected behaviour, it's not like we're going to be able to get them to change it. So we should store preTLS, preAuth and postAuth capabilities. We should assume subsequent capabilities are the same as the previous unless replaced.

**Comment by philipwhiuk** [comment]
In fact, my IMAP mail server does this so I'm assuming it is a core part of IMAP's behaviour. Anyway, it's not so hard to fix.

**Comment by philipwhiuk** [comment]
It occurs to me that maybe this only occurs because we now properly close InputStreams so we actually can re-use an ImapConnection object that got closed. (It's a wild guess, but it might be the case).

**Comment by cketti** [comment]
Right now `ImapConnection` isn't designed to be reused. And I'd rather not change that. So instead of having two caches for capabilities we should make sure `ImapConnection` is not reused.

**Comment by philipwhiuk** [comment]
I'm not sure that's true. Why would send command call open(). You'd only be able to open it once and then it would be a no-op.

Here's the stacktrace I got while testing 

```
com.fsck.k9.mail.MessagingException: Server doesn't support <<CRYPTO METHOD>>.
at com.fsck.k9.mail.store.imap.ImapConnection.authenticate(ImapConnection.java:328)
at com.fsck.k9.mail.store.imap.ImapConnection.open(ImapConnection.java:120)
at com.fsck.k9.mail.store.imap.ImapConnection.sendCommand(ImapConnection.java:696)
at com.fsck.k9.mail.store.imap.ImapConnection.executeSimpleCommand(ImapConnection.java:654)
at com.fsck.k9.mail.store.imap.ImapConnection.executeSimpleCommand(ImapConnection.java:643)
at com.fsck.k9.mail.store.imap.ImapFolder.executeSimpleCommand(ImapFolder.java:109)
at com.fsck.k9.mail.store.imap.ImapFolder.internalOpen(ImapFolder.java:126)
at com.fsck.k9.mail.store.imap.ImapFolderPusher$PushRunnable.openConnectionIfNecessary(ImapFolderPusher.java:328)
at com.fsck.k9.mail.store.imap.ImapFolderPusher$PushRunnable.run
```

Looking at the code, the very name 'openConnectionIfNecessary' implies to me it expects that the connection can close and be re-opened.

**Comment by cketti** [comment]
`ImapFolderPusher` only reuses connections that haven't been closed. You can see that e.g. `ImapFolder.close()` also discards the `ImapConnection` instance when it closes the connection.

It's unclear what the initial goal was for `ImapConnection`. But my refactoring makes the assumption that `ImapConnection` is not reused once it has been closed. That's mainly to avoid having to deal with leftover state from a previous connection.

**Comment by philipwhiuk** [comment]
Reproduced the issue with the fix. Here's the trace. 


```
java.lang.IllegalStateException: open() called after close(). Check wrapped exception to see where close() was called.
at com.fsck.k9.mail.store.imap.ImapConnection.open(ImapConnection.java:107)
at com.fsck.k9.mail.store.imap.ImapConnection.sendCommand(ImapConnection.java:718)
at com.fsck.k9.mail.store.imap.ImapConnection.executeSimpleCommand(ImapConnection.java:676)
at com.fsck.k9.mail.store.imap.ImapConnection.executeSimpleCommand(ImapConnection.java:665)
at com.fsck.k9.mail.store.imap.ImapFolder.executeSimpleCommand(ImapFolder.java:109)
at com.fsck.k9.mail.store.imap.ImapFolder.internalOpen(ImapFolder.java:126)
at com.fsck.k9.mail.store.imap.ImapFolderPusher$PushRunnable.openConnectionIfNecessary(ImapFolderPusher.java:328)
at com.fsck.k9.mail.store.imap.ImapFolderPusher$PushRunnable.run(ImapFolderPusher.java:155)
at java.lang.Thread.run(Thread.java:818)
Caused by: java.lang.Exception
at com.fsck.k9.mail.store.imap.ImapConnection.close(ImapConnection.java:645)
at com.fsck.k9.mail.store.imap.ImapFolderPusher$PushRunnable.reacquireWakeLockAndCleanUp(ImapFolderPusher.java:252)
at com.fsck.k9.mail.store.imap.ImapFolderPusher$PushRunnable.run(ImapFolderPusher.java:202)
... 1 more
```

**Comment by cketti** [comment]
Hopefully fixed by #1348
